File: frontend/src/utils/actionEligibility.ts
/**
 * Quick-action eligibility mirror of backend action_eligibility.py.
 * Keep reason codes and messages aligned with the Python SoT.
 */
import type { LeadStatus } from '@/types'
import { formatDateOnly } from '@/utils/helpers'

export const REASON_ALREADY_SKIP_TRACE = 'already_skip_trace'
export const REASON_ALREADY_AWAITING_SKIP_TRACE = 'already_awaiting_skip_trace'
export const REASON_TERMINAL_STATUS = 'terminal_status'
export const REASON_DNC_BLOCKS_OUTREACH = 'dnc_blocks_outreach'
export const REASON_MAIL_RECENTLY_SOLD = 'mail_recently_sold'
export const REASON_MAIL_INVALID_ADDRESS = 'mail_invalid_address'
export const REASON_MAIL_ALREADY_QUEUED = 'mail_already_queued'

const TERMINAL_LEAD_STATUSES: LeadStatus[] = [
  'deprioritize',
  'deal_won',
  'deal_lost',
  'suppressed',
  'do_not_contact',
]

export type QuickActionId =
  | 'log_call'
  | 'log_note'
  | 'log_email'
  | 'add_to_mail_batch'
  | 'move_to_skip_trace'

export interface ActionEligibilityResult {
  ok: boolean
  reasonCode: string | null
  message: string | null
  alreadyDone: boolean
}

function ok(): ActionEligibilityResult {
  return { ok: true, reasonCode: null, message: null, alreadyDone: false }
}

function blocked(
  reasonCode: string,
  message: string,
  alreadyDone = false,
): ActionEligibilityResult {
  return { ok: false, reasonCode, message, alreadyDone }
}

export function evaluateMoveToSkipTrace(leadStatus: LeadStatus): ActionEligibilityResult {
  if (leadStatus === 'skip_trace') {
    return blocked(REASON_ALREADY_SKIP_TRACE, 'Already in Skip Trace', true)
  }
  if (leadStatus === 'awaiting_skip_trace') {
    return blocked(REASON_ALREADY_AWAITING_SKIP_TRACE, 'Already awaiting skip trace', true)
  }
  if (TERMINAL_LEAD_STATUSES.includes(leadStatus)) {
    return blocked(REASON_TERMINAL_STATUS, 'Not available for this lead status')
  }
  return ok()
}

export function evaluateAddToMailBatch(input: {
  mailQueueStatus?: 'queued' | 'sent_recently' | null
  mailEligible?: boolean
  isMailable?: boolean
  mailIneligibleReason?: string | null
  mailEligibleDate?: string | null
}): ActionEligibilityResult {
  if (input.mailQueueStatus === 'queued') {
    return blocked(
      REASON_MAIL_ALREADY_QUEUED,
      'Already staged for the next mail batch',
      true,
    )
  }
  const mailReady = input.mailEligible ?? input.isMailable ?? false
  if (mailReady) return ok()
  if (input.mailIneligibleReason === 'recently_sold') {
    if (input.mailEligibleDate) {
      return blocked(
        REASON_MAIL_RECENTLY_SOLD,
        `Held after recent sale until ${formatDateOnly(input.mailEligibleDate)}`,
      )
    }
    return blocked(
      REASON_MAIL_RECENTLY_SOLD,
      'Held after recent sale until the two-year hold ends',
    )
  }
  return blocked(
    REASON_MAIL_INVALID_ADDRESS,
    'Owner mailing address is not ready for the mail queue',
  )
}

export function evaluateOutreachLog(
  leadStatus: LeadStatus,
  action: 'log_call' | 'log_note' | 'log_email',
): ActionEligibilityResult {
  if (action === 'log_note') return ok()
  if (
    (action === 'log_call' || action === 'log_email')
    && leadStatus === 'do_not_contact'
  ) {
    return blocked(
      REASON_DNC_BLOCKS_OUTREACH,
      'Outreach is blocked — lead is Do Not Contact',
    )
  }
  return ok()
}

/** Unavailable title text for a Quick Action button, or null if enabled. */
export function unavailableReasonForQuickAction(
  action: QuickActionId,
  ctx: {
    leadStatus: LeadStatus
    mailQueueStatus?: 'queued' | 'sent_recently' | null
    mailEligible?: boolean
    isMailable?: boolean
    mailIneligibleReason?: string | null
    mailEligibleDate?: string | null
  },
): string | null {
  if (action === 'move_to_skip_trace') {
    const r = evaluateMoveToSkipTrace(ctx.leadStatus)
    return r.ok ? null : r.message
  }
  if (action === 'add_to_mail_batch') {
    const r = evaluateAddToMailBatch(ctx)
    // Queued shows alternate "In mail batch" controls — not a gray Add button title.
    if (r.reasonCode === REASON_MAIL_ALREADY_QUEUED) return null
    return r.ok ? null : r.message
  }
  if (action === 'log_call' || action === 'log_email' || action === 'log_note') {
    const r = evaluateOutreachLog(ctx.leadStatus, action)
    return r.ok ? null : r.message
  }
  return null
}

File: backend/app/services/action_eligibility.py
"""Quick-action eligibility — source of truth for Command Center Quick Actions.

Frontend mirrors reason codes in ``frontend/src/utils/actionEligibility.ts``.
Keep both in sync when adding actions or reason codes.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from typing import Optional, Protocol


TERMINAL_LEAD_STATUSES = frozenset({
    'deprioritize',
    'deal_won',
    'deal_lost',
    'suppressed',
    'do_not_contact',
})

SKIP_TRACE_PIPELINE_STATUSES = frozenset({
    'skip_trace',
    'awaiting_skip_trace',
})

# Stable reason codes — mirrored in frontend actionEligibility.ts
REASON_ALREADY_SKIP_TRACE = 'already_skip_trace'
REASON_ALREADY_AWAITING_SKIP_TRACE = 'already_awaiting_skip_trace'
REASON_TERMINAL_STATUS = 'terminal_status'
REASON_DNC_BLOCKS_OUTREACH = 'dnc_blocks_outreach'
REASON_MAIL_RECENTLY_SOLD = 'mail_recently_sold'
REASON_MAIL_INVALID_ADDRESS = 'mail_invalid_address'
REASON_MAIL_ALREADY_QUEUED = 'mail_already_queued'


class _LeadStatusLike(Protocol):
    lead_status: Optional[str]


@dataclass(frozen=True)
class ActionEligibilityResult:
    ok: bool
    reason_code: Optional[str] = None
    message: Optional[str] = None
    already_done: bool = False


def _ok() -> ActionEligibilityResult:
    return ActionEligibilityResult(ok=True)


def _blocked(
    reason_code: str,
    message: str,
    *,
    already_done: bool = False,
) -> ActionEligibilityResult:
    return ActionEligibilityResult(
        ok=False,
        reason_code=reason_code,
        message=message,
        already_done=already_done,
    )


def evaluate_move_to_skip_trace(lead: _LeadStatusLike) -> ActionEligibilityResult:
    """Whether Move to Skip Trace may mutate this lead.

    ``skip_trace`` is already on the Skip Trace work column (already done).
    ``awaiting_skip_trace`` (e.g. recent-sale hold ended) still needs an active
    handoff into that queue — Move to Skip Trace must stay available.
    """
    status = lead.lead_status
    if status == 'skip_trace':
        return _blocked(
            REASON_ALREADY_SKIP_TRACE,
            'Already in Skip Trace',
            already_done=True,
        )
    if status in TERMINAL_LEAD_STATUSES:
        return _blocked(
            REASON_TERMINAL_STATUS,
            'Not available for this lead status',
        )
    return _ok()


def evaluate_add_to_mail_batch(
    *,
    mail_queue_status: Optional[str] = None,
    mail_eligible: bool = False,
    mail_ineligible_reason: Optional[str] = None,
    mail_eligible_date: Optional[date | str] = None,
) -> ActionEligibilityResult:
    """Whether Add to Mail Queue should be enabled (UI) / treated as ready.

    Soft enqueue still returns per-lead outcomes; this gates the Quick Action.
    """
    if mail_queue_status == 'queued':
        return _blocked(
            REASON_MAIL_ALREADY_QUEUED,
            'Already staged for the next mail batch',
            already_done=True,
        )
    if mail_eligible:
        return _ok()
    if mail_ineligible_reason == 'recently_sold':
        if mail_eligible_date:
            if isinstance(mail_eligible_date, date):
                when = mail_eligible_date.isoformat()
            else:
                when = str(mail_eligible_date)
            return _blocked(
                REASON_MAIL_RECENTLY_SOLD,
                f'Held after recent sale until {when}',
            )
        return _blocked(
            REASON_MAIL_RECENTLY_SOLD,
            'Held after recent sale until the two-year hold ends',
        )
    return _blocked(
        REASON_MAIL_INVALID_ADDRESS,
        'Owner mailing address is not ready for the mail queue',
    )


def evaluate_outreach_log(
    lead: _LeadStatusLike,
    action: str,
) -> ActionEligibilityResult:
    """Log Call / Log Email blocked on DNC; Log Note always ok."""
    if action == 'log_note':
        return _ok()
    if action in {'log_call', 'log_email'} and lead.lead_status == 'do_not_contact':
        return _blocked(
            REASON_DNC_BLOCKS_OUTREACH,
            'Outreach is blocked — lead is Do Not Contact',
        )
    return _ok()
